File: day15/dueling_generators2_iter.cpp
#include <iostream>
#include <vector>
#include "array2d.h"
using namespace std;

class DuelGen {
private:
    unsigned long _val;
    unsigned long _fact;
    unsigned long _mod;

    void genNextVal() { _val = (_val * _fact) % 2147483647; }

public:
    DuelGen(const unsigned long startVal, const unsigned long factor, const unsigned long mod) {
        _val = startVal;
        _fact = factor;
        _mod = mod;
    }

    unsigned long next() {
        do {
            genNextVal();
        } while (_val % _mod != 0);
        return _val;
    }
};

int main(int argc, char *argv[]) {
//    unsigned long gen[] = {65, 8921};
    unsigned long startVal[] = {116, 299};
    const unsigned long FACT[] = {16807, 48271};
    const unsigned long MOD[] = {4, 8};
    vector<DuelGen> gen;
//    const int PAIRS = 5;
    const int PAIRS = 5000000;

    for (int i = 0; i < 2; i++)
        gen.push_back(DuelGen(startVal[i], FACT[i], MOD[i]));
        
    int cnt = 0;
    for (int i = 0; i < PAIRS; i++)
        if ((gen[0].next() & 0xffff) == (gen[1].next() & 0xffff))
            cnt++;

    cout << "result2: " << cnt << endl;
}

File: day15/dueling_generators2_iter.py
#!/usr/bin/env python3

class DuelGen:
    """Class to implement an iterator for the generators in part 2
    of Advent of Code Day 15"""

    def __init__(self, startVal, factor, mod):
        self.val = startVal
        self.fact = factor
        self.mod = mod

    def __iter__(self):
        return self

    def __next__(self):
        self.val = (self.val * self.fact) % 2147483647
        while self.val % self.mod != 0:
            self.val = (self.val * self.fact) % 2147483647
        return self.val

# startVal = [65, 8921]; # test
startVal = [116, 299];
FACT = [16807, 48271]
MOD = [4, 8]
PAIRS = 5000000
gen = [iter(DuelGen(startVal[i], FACT[i], MOD[i])) for i in range(2)]
cnt = 0

for i in range(PAIRS):
    if (next(gen[0]) & 0xffff) == (next(gen[1]) & 0xffff):
        cnt += 1

print("result2:", cnt)


